alphabase/constants/atom.py

```diff
@@ -200,3 +204,77 @@ def calc_mass_from_formula(formula: str):
         mass of the formula
     """
     return np.sum([CHEM_MONO_MASS[elem] * n for elem, n in parse_formula(formula)])
+
+
+class ChemicalCompositonFormula:
+    def __init__(self, formula):
+        self.elements = self._parse_formula(formula)
+
+    @classmethod
+    def from_rdkit_mol(cls, mol: Chem.Mol):
+        if not mol:
+            raise ValueError("Invalid RDKit molecule")
+        formula = rdMolDescriptors.CalcMolFormula(
+            mol, separateIsotopes=True, abbreviateHIsotopes=False
+        )
+        formula = formula.replace("[1H]", "H")
+        return cls._from_rdkit_formula(formula)
+
+    @classmethod
+    def _from_rdkit_formula(cls, formula):
+        instance = cls.__new__(cls)
+        instance.elements = instance._parse_rdkit_formula(formula)
+        return instance
+
+    def _parse_formula(self, formula):
+        pattern = r"(\d+)?([A-Z][a-z]*)(?:\(([-]?\d+)\))?"
+        matches = re.findall(pattern, formula)
+        element_counts = defaultdict(int)
+
+        for isotope, element, count in matches:
+            if isotope:
+                element = f"{isotope}{element}"
+            count = int(count) if count else 1
+            element_counts[element] += count
+
+        return element_counts
+
+    def _parse_rdkit_formula(self, formula):
+        pattern = r"(\[(\d+)([A-Z][a-z]*)\]|([A-Z][a-z]*))(\d*)"
+        matches = re.findall(pattern, formula)
+        element_counts = defaultdict(int)
+
+        for match in matches:
+            if match[1]:  # Isotope
```

Comment:
mb give a short comment how `match` looks like in the two cases

Reply:
Done!